frontend/src/Login.tsx

```diff
@@ -0,0 +1,84 @@
+import React from "react";
+import { EventBus, EventBusEventType } from "./EventBus";
+import { LoginManager } from "./LoginManager";
+import { Redirect } from "react-router-dom";
+
+export interface LoginCredentials {
+    username: string
+    password: string
+    eventBus: EventBus
+    loginManager: LoginManager
+    loggedIn?: boolean
+}
+
+/** Login Page. Consult the LoginManger.tsx for further login logic. */
+export class Login extends React.Component<LoginCredentials, LoginCredentials> {
+    constructor(props: LoginCredentials, state: LoginCredentials) {
+        super(props);
+        this.state = {username: props.username, password: props.password, eventBus: props.eventBus, loginManager: props.loginManager, loggedIn: props.loginManager.isLoggedIn()};
+        props.eventBus.register(EventBusEventType.USER_CHANGE, (eventType, eventData?) => this.onUserChange(eventData));
+    }
+
+    private onUserChange(eventData?: any) {
+        this.setState({username: this.state.username, password: this.state.password,
+            eventBus: this.state.eventBus,
+            loginManager: this.state.loginManager, loggedIn: eventData.loggedIn});
+    }
+
+    private formInputFieldUpdated(event: React.ChangeEvent<HTMLInputElement>) {
+        this.update(event.currentTarget.id, event.currentTarget.value);
+    }
+
+    private update(id: string, value: string) {
+        if (id === "username") {
+            this.setState({
+                username: value,
+                password: this.state.password,
+                eventBus: this.state.eventBus,
+                loginManager: this.state.loginManager,
+                loggedIn: this.state.loggedIn
+            });
+        }
+        if (id === "password") {
+            this.setState({
+                username: this.state.username,
+                password: value,
+                eventBus: this.state.eventBus,
+                loginManager: this.state.loginManager,
+                loggedIn: this.state.loggedIn
+            });
+        }
+    }
```

Comment:
Maybe simplify it like this:
```suggestion
     private update(id: string, value: string) {
        let newState = {
            username: this.state.username,
            password: this.state.password,
            eventBus: this.state.eventBus,
            loginManager: this.state.loginManager,
            loggedIn: this.state.loggedIn
        };
        if (newState[id] != null) {
            newState[id] = value;
        }
        this.setState(newState);
    }
```

Reply:
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ username: string; password: string; eventBus: EventBus; loginManager: LoginManager; loggedIn: boolean | undefined; }'.
  No index signature with a parameter of type 'string' was found on type '{ username: string; password: string; eventBus: EventBus; loginManager: LoginManager; loggedIn: boolean | undefined; }'.

Reply:
Does `newState[id as string]` solve it?